File: FoosballGames/TypeMappers.cs
using FoosballGames.Infrastructure;

namespace FoosballGames;

public static class TypeMappers
{
    public static Contracts.FoosballGame ToContract(this FoosballGame game)
    {
        return new(
            Id: game.Id,
            Start: game.Start.ToDateTimeUnspecified(),
            End: game.End?.ToDateTimeUnspecified(),
            Sets: game.Sets.ToContract().ToArray(),
            Finished: game.Sets is FinishedSets,
            WinnerTeam: game.Sets is FinishedSets finishedSets ? finishedSets.WinnerTeam : null
        );
    }

    public static IEnumerable<Contracts.Set> ToContract(this ISets sets)
    {
        switch (sets)
        {
            case FinishedSets finishedSets:
                var response = finishedSets.FirstSet
                    .ToContract(1)
                    .ToSingletonEnumerable()
                    .Append(finishedSets.SecondSet.ToContract(2));
                if (finishedSets.ThirdSet is not null)
                    response = response.Append(finishedSets.ThirdSet.ToContract(3));
                return response;
            case FirstSetRunning firstSetRunning:
                return firstSetRunning.Set.ToContract(1).ToSingletonEnumerable();
            case SecondSetRunning secondSetRunning:
                return secondSetRunning.FirstSet
                    .ToContract(1)
                    .ToSingletonEnumerable()
                    .Append(secondSetRunning.SecondSet.ToContract(2));
            case ThirdSetRunning thirdSetRunning:
                return thirdSetRunning.FirstSet
                    .ToContract(1)
                    .ToSingletonEnumerable()
                    .Append(thirdSetRunning.SecondSet.ToContract(2))
                    .Append(thirdSetRunning.ThirdSet.ToContract(3));
            default:
                throw new ArgumentOutOfRangeException(nameof(response));
        }
    }

    public static Contracts.Set ToContract(this ISet set, int number)
    {
        return set switch
        {
            FinishedSet finishedSet => new Contracts.Set(number, true, RedTeamScore: finishedSet.RedTeamScore,
                BlueTeamScore: finishedSet.BlueTeamScore),
            RunningSet runningSet => new Contracts.Set(number, false, RedTeamScore: runningSet.RedTeamScore,
                BlueTeamScore: runningSet.BlueTeamScore),
            _ => throw new ArgumentOutOfRangeException(nameof(set))
        };
    }
}

File: FoosballGames/TypeMappers.fs
module FoosballGames.TypeMappers

open System
open FoosballGames

type ISet with
    member this.ToContract number : Contracts.Set =
        match this with
        | :? FinishedSet as f ->
            { Number = number; Finished = true; RedTeamScore = f.RedTeamScore; BlueTeamScore = f.BlueTeamScore }
        | :? RunningSet as r ->
            { Number = number; Finished = false; RedTeamScore = r.RedTeamScore; BlueTeamScore = r.BlueTeamScore }
        | _ -> raise <| ArgumentOutOfRangeException("Set", this, "Unknown set type")


type ISets with
    member this.ToContract() : Contracts.Set seq =
        match this with
        | :? FinishedSets as f ->
            seq {
                yield f.FirstSet.ToContract 1
                yield f.SecondSet.ToContract 2
                match f.ThirdSet with
                | Some s -> yield s.ToContract 3
                | None -> ()
            }
        | :? FirstSetRunning as f ->
            f.Set.ToContract 1 |> Seq.singleton
        | :? SecondSetRunning as s ->
            seq {
                yield s.FirstSet.ToContract 1
                yield s.SecondSet.ToContract 2
            }
        | :? ThirdSetRunning as t ->
            seq {
                yield t.FirstSet.ToContract 1
                yield t.SecondSet.ToContract 2
                yield t.ThirdSet.ToContract 3
            }
        | _ -> raise <| ArgumentOutOfRangeException("Sets", this, "Unknown sets type")

type FoosballGame with
    member this.ToContract() : Contracts.FoosballGame =
        { Id = this.Id
          Start = this.Start.ToDateTimeUnspecified()
          End = this.End
                |> Option.map (fun x -> x.ToDateTimeUnspecified())
                |> Option.toNullable
          Sets = this.Sets.ToContract() |> Seq.toArray
          Finished = this.Sets :? FinishedSets
          WinnerTeam =
                match this.Sets with
                | :? FinishedSets as f -> Nullable f.WinnerTeam
                | _ -> Nullable()
           }
